### PR TITLE
feat: support torch.optim.lr_scheduler.ReduceLROnPlateau

### PR DESCRIPTION
**Pull request by brain-good**
# Test Plan
- UI change:
  - [ ] add screenshots
  - [ ] React? build & check storybooks
- [ ] user-facing api change: modify documentation and examples
- [ ] user-facing api change: add the "User-facing API Change" label
- [ ] bug fix: add regression test
- [ ] bug fix: determine if there are other similar bugs in the codebase
- [ ] new feature: add test coverage for any user-facing aspects
- [ ] refactor: maintain existing code coverage

### CODE REVIEW

File: harness/determined/pytorch/_lr_scheduler.py [hunk 1]
```diff
@@ -5,6 +5,10 @@
 
 import determined_common.check as check
 
+TorchLRScheduler = typing.Union[
```
Comment (aaron276h) [inline harness/determined/pytorch/_lr_scheduler.py:8]: ? expose this in `__init__` since user facing

Comment (brain-good) [reply harness/determined/pytorch/_lr_scheduler.py:8]: it's not really user facing though right? We just use this internally here to make mypy happy

Comment (aaron276h) [reply harness/determined/pytorch/_lr_scheduler.py:8]: I think input types for user facing functions/classes (such as `LRScheduler`)  should be considered user facing.

**Comment by brain-good** [comment]
> It's pretty likely that users wile use step() in a way that would lead to incorrect behavior in multi-GPU training, we can support it correctly by synching the across the workers. One approach we could take is to offer our own torch.optim.lr_scheduler.ReduceLROnPlateau which subclasses the original class and overrides the step() function to sync the state.

I agree that this can lead to incorrect behavior in multi-GPU case, but had something else in mind for our support for that. I think we will get a much cleaner solution to this scenario after we implement "Better support for PyTorch callbacks" as we discussed in the last ML AG. Users could call ReduceLROnPlateau as part of a "post validation callback". This is better for the single-GPU case since users may have implemented `validate_batch` in which case they don't have access to the complete validation metric. Maybe we make this change but don't officially document our support yet until we do the callbacks. Let me know what you think @aaron276h

File: harness/determined/pytorch/_lr_scheduler.py (hunk 1 above)
Comment (brain-good) [reply harness/determined/pytorch/_lr_scheduler.py:8]: I personally don't really see the rationale. They're going to import it and do what with it? If you think it's important and it's something we've done elsewhere then I'll do it

Comment (brain-good) [reply harness/determined/pytorch/_lr_scheduler.py:8]: It seems like we do that for types that users will have to implement a method for (`TorchData`). I'll do it for this one anyways 👍

**Comment by aaron276h** [comment]
@brain-good Yes having the callbacks will help, but they will solve the issue of "when is step called" rather than the problem keeping the state in sync across all the workers (unless you would sync validation stats across all the workers prior the the callback, and even then I think relying on just the callback has the drawback of being very restrictive about when `step()` can be called).

File: harness/determined/pytorch/_lr_scheduler.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/pytorch/_lr_scheduler.py:8]: `keras.InputData` is a similar thing although it's an expected output vs. arg. But yeah I do agree with you, we should just discuss this in #ml-ag rather than trying to figure out the standard in this thread.

**Comment by brain-good** [comment]
> unless you would sync validation stats across all the workers prior the the callback, and even then I think relying on just the callback has the drawback of being very restrictive about when step() can be called

Yes, I definitely think a callback related to validation would be necessary to be post-sync of the validation stats. Since a usecase for that would be this exact thing I'm talking about. In terms of being restrictive of when this could be called, I think it's important to keep in mind what this callback is actually used for: `Reduce learning rate when a metric has stopped improving.`. It is used to reduce the learning rate based on a validation metric. I don't see a reason to use this at any point but after sync validation metrics.

I think that what we both have in mind are pretty far from each other. Probably makes sense to discuss over Zoom on Monday.

**Comment by brain-good** [comment]
After some discussion, we've decided to do what @aaron276h suggested in the short-term, and consider a different approach in the future when we have general support for training hooks
